pkg/sockops/sockops.go

```diff
@@ -0,0 +1,491 @@
+	log.WithFields(logrus.Fields{
+		"bpftool": prog,
+		"args":    args,
+	}).Debug("GetMapID:")
+	output, err := exec.Command(prog, args...).CombinedOutput()
+	if err != nil {
+		return 0, fmt.Errorf("Failed to load %s: %s", progName, err)
+	}
+
+	// Find the mapID out of the bpftool output
+	s := strings.Fields(string(output))
+	for i := range s {
+		if s[i] == "map_ids" {
+			id := strings.Split(s[i+1], ",")
+			for j := range id {
+				args := []string{"map", "show", "id", id[j]}
+				output, err := exec.Command(prog, args...).CombinedOutput()
+				if err != nil {
+					return 0, err
+				}
+
+				if strings.Contains(string(output), mapName) {
+					mapID, _ := strconv.Atoi(id[j])
+					return mapID, nil
+				}
+			}
+			break
+		}
+	}
+	return 0, nil
+}
+
+// #bpftool map pin id map_id /sys/fs/bpf/tc/globals
+func bpftoolPinMapID(mapName string, mapID int) error {
+	bpffs := bpf.GetMapRoot()
+	globals := bpffs + "/" + mapPrefix + "/"
+	mapFile := globals + mapName
+	prog := "bpftool"
+
+	args := []string{"map", "pin", "id", strconv.Itoa(mapID), mapFile}
+	log.WithFields(logrus.Fields{
+		"bpftool": prog,
+		"args":    args,
+	}).Debug("Map pin:")
+	_, err := exec.Command(prog, args...).CombinedOutput()
+	if err != nil {
+		return fmt.Errorf("Failed to pin map %d(%s): %s", mapID, mapName, err)
+	}
+
+	return nil
+}
+
+// #clang ... | llc ...
+func bpfCompileProg(src string, dst string) error {
+	ctx, cancel := context.WithTimeout(context.Background(), contextTimeout)
+	defer cancel()
+
+	srcpath := filepath.Join("sockops", src)
+	outpath := filepath.Join(dst)
+
+	err := loader.Compile(ctx, srcpath, outpath)
+	if err != nil {
+		return fmt.Errorf("failed compile %s: %s", srcpath, err)
+	}
+	return nil
+}
+
+func bpfLoadMapProg(object string, load string) error {
+	sockops := object
+	sockopsObj := option.Config.StateDir + "/" + sockops
+	sockopsLoad := load
+
+	err := bpftoolLoad(sockopsObj, sockopsLoad)
+	if err != nil {
+		return err
+	}
+
+	progID, err := bpftoolGetProgID(load)
+	if err != nil {
+		return err
+	}
+
+	_mapID, err := bpftoolGetMapID("bpf_sockops", sockMap)
+	mapID := strconv.Itoa(_mapID)
+	if err != nil {
+		return err
+	}
+
+	err = bpftoolMapAttach(progID, mapID)
+	if err != nil {
+		return err
+	}
+	skmsgLockedDown = true
+	return nil
+}
+
+// SkmsgEnable will compile and attach the SK_MSG programs to the
+// sockmap. After this all sockets added to the sock_ops_map will
+// have sendmsg/sendfile calls running through BPF program.
+func SkmsgEnable() error {
+	err := bpfCompileProg(cIPC, oIPC)
+	if err != nil {
+		log.Error(err)
+		return err
+	}
+
+	err = bpfLoadMapProg(oIPC, eIPC)
+	if err != nil {
+		log.Error(err)
+		return err
+	}
+	log.Infof("Sockmsg Enabled, bpf_redir loaded")
+	return nil
+}
+
+// SkmsgDisable "unloads" the SK_MSG program. This simply deletes
+// the file associated with the program.
+func SkmsgDisable() {
+	bpftoolUnload(eIPC)
+	log.Infof("Sockmsg Disabled.")
```

Comment:
Should this be kept as info? Also should be `Info` not `Infof`

Reply:
I have no strong preference, would debug be more appropriate?